app/services/spotify_service.rb

```diff
@@ -1,84 +1,83 @@
-
-        # Attach new artists to the current user
-        @current_user.followed_artists.create!(artist_ids.map { |artist_id| { artist_id: artist_id } })
-      end
+  # This method makes a request to the Spotify API to fetch the total number of artists followed by the current user.
+  #
+  # @return [Integer] The total number of artists followed by the current user.
+  def total_followed_artists
+    RSpotify.raw_response = true
+    count = JSON.parse(@spotify_user.following(type: 'artist'))["artists"]["total"]
+    RSpotify.raw_response = false
+
+    count
+  rescue => e
+    Rails.logger.error("Unexpected error when fetching total followed artists from Spotify: #{e.message}")
+  end
 
-      remove_unfollowed_artists(all_fetched_artists, followed_artist_names)
+  # This method fetches all followed artists of the current user from Spotify and converts them into an array of Artist
+  # objects.
+  #
+  # @return [Array<Artist>] An array containing all fetched artists.
+  def artists
+    fetch_all_followed_artists.map do |a|
+      Artist.new(
+        name: a.name,
+        external_link: a.uri,
+        cover_url: a.images.last&.dig("url")
+      )
     end
   end
 
   private
 
-  # Use Spotify's API to fetch the following artists
-  def fetch_artists(last_artist_id)
-    @current_user.spotify_user.following(type: 'artist', limit: 50, after: last_artist_id)
-  end
-
-  # Removes unfollowed artists from the current user's followed artists.
+  # Fetches all followed artists of the current user from Spotify.
   #
-  # @param all_followed_artists [Array<Array<Artist>>] the list of all followed artists fetched from Spotify
-  # @param previously_followed_artist_names [Array<String>] the list of names of artists previously followed by the user
-  # @return [void]
-  def remove_unfollowed_artists(all_followed_artists, previously_followed_artist_names)
-    followed_artists_names = all_followed_artists.flatten.map(&:name)
-    unfollowed_artists_names = previously_followed_artist_names - followed_artists_names
-
-    return if unfollowed_artists_names.empty?
+  # This method retrieves all followed artists of the current user from Spotify in batches until either the total number
+  # of fetched artists reaches the total number of followed artists or the maximum loop count is exceeded.
+  #
+  # @return [Array<Artist>] An array containing all fetched artists.
+  def fetch_all_followed_artists
+    artists = []
+    count_loop = 0
+
+    loop do
+      count_loop += 1
+      batch_of_fetched_artists = fetch_batch_of_followed_artists(artists.last&.id)
+      artists.concat(batch_of_fetched_artists)
+
+      # Break the loop if:
+      # - The batch of fetched artists is empty.
+      # - The total number of fetched artists exceeds or equals the total number of followed artists.
+      # - The maximum loop count exceeds 100.
+      break if batch_of_fetched_artists.empty? || artists.size >= total_followed_artists || count_loop > MAX_LOOP
```

Comment:
Attention ici tu stop ton fetch dès que tu excéde 10 itérations.
Vu que tu récupère le nombre total d'artistes à récupérer du pourrais calculer le nombre d'itérations nécéssaires pour tout fetch.

```suggestion
      # - The maximum loop count exceeds 100.
      
      artists_to_fetch = total_followed_artists
      iterations_needed = artists_to_fetch / SPOTIFY_MAX_LIMIT_PER_PAGE
      break if batch_of_fetched_artists.empty? || count_loop > iterations_needed
```

Reply:
Bien vu ! J'ai modifié cette ligne pendant mes tests, j'ai zappé de la remettre avec un nombre élevé comme 100.

> Vu que tu récupère le nombre total d'artistes à récupérer tu pourrais calculer le nombre d'itérations nécessaires pour tout fetch.

J'ai modifié mon code avec ta proposition, est-ce que cela te convient 😊 ?

Reply:
@davidBentoPereira C'est vachement mieux !
Dernier petit détail:
Tu appelle `total_followed_artists` à chaque itération, ce qui fait un call supplémentaire à Spotify.
La donnée renvoyée ne devrait pas changer entre 2 appels et ça ralentit ton service.

Tu n'as normalement plus besoin de checker cette valeur systématiquement vu que tu calcule déjà le nombre de tours de boucle max :)

Reply:
Oh ! Bien vu ! 🙏